src/platform/linker/loadall.c

```diff
@@ -20,27 +20,53 @@
 #include <assert.h>
 #include <sys/mman.h>
 
+char *service_name = NULL;
+
+struct private_symtab{
+	char name[MAX_SYMB_LEN];
+	unsigned long vma_val;
+};
+
 unsigned long getsym(bfd *obj, char* symbol)
 {
         long storage_needed;
-        asymbol **symbol_table;
+        asymbol **symbol_table = NULL;
+	static struct private_symtab *private_symtabs;
+	static char prev_name[MAX_FILE_NAME_LEN] = "";
         long number_of_symbols;
         int i;
 
-        storage_needed = bfd_get_symtab_upper_bound (obj);
+	assert(service_name);
+	if (strcmp(prev_name, service_name)) {
+		strcpy(prev_name, service_name);
 
-        if (storage_needed <= 0){
-                printl(PRINT_HIGH, "no symbols in object file\n");
-                exit(-1);
-        }
+		storage_needed = bfd_get_symtab_upper_bound (obj);
+		printl(PRINT_DEBUG, "\tAllocating new symbol table...\n");
 
-        symbol_table = (asymbol **) malloc (storage_needed);
-        number_of_symbols = bfd_canonicalize_symtab(obj, symbol_table);
+		if (storage_needed <= 0){
+		        printl(PRINT_HIGH, "no symbols in object file\n");
+		        exit(-1);
+		}
+
+		symbol_table = (asymbol **) malloc (storage_needed);
+		assert(symbol_table);
+		number_of_symbols = bfd_canonicalize_symtab(obj, symbol_table);
+		private_symtabs = malloc (number_of_symbols * sizeof(struct private_symtab));
+		assert(private_symtabs);
+		int i;
+		for (i = 0 ; i < number_of_symbols ; i++) {
+			assert(symbol_table[i]->name);
+			strcpy(private_symtabs[i].name, symbol_table[i]->name);
+			private_symtabs[i].vma_val = symbol_table[i]->section->vma + symbol_table[i]->value;
+		}
+	}
+
+	assert(private_symtabs != NULL);
 
         //notes: symbol_table[i]->flags & (BSF_FUNCTION | BSF_GLOBAL)
         for (i = 0; i < number_of_symbols; i++) {
-                if(!strcmp(symbol, symbol_table[i]->name)){
-                        return symbol_table[i]->section->vma + symbol_table[i]->value;
+                if(!strcmp(symbol,  private_symtabs[i].name)){
+                        return private_symtabs[i].vma_val;
```

Comment:
Why can't we just free `private_symtabs` after this?

Reply:
Yes and no. We should be able to free private_symtabs, but not here. If we free here we would possibly crash next time this function is executed if it expected the cached private_symtabs to be there. We could cache vma_val, but that is adding complexity through additional variables. I added a free just prior to allocating the next private_symtabs and tested with llbooter_pong.sh and it worked fine. Can't test as of right now with rumpkernboot.sh, though.